### ISSUE TITLE
Ignition support for Vagrant CoreOS

### ISSUE THREAD

**Issue by larstobi** [coreos/bugs#1891]
# Issue Report #

### Container Linux Version ###

```
$ cat /etc/os-release
NAME="Container Linux by CoreOS"
...
BUG_REPORT_URL="https://github.com/coreos/bugs/issues"
```

### Environment ###

What hardware/cloud provider/hypervisor is being used to run Container Linux?

Vagrant and VMware Fusion.

### Expected Behavior ###

Vagrant reads Ignition info into user-data and boots.

### Actual Behavior ###

Vagrant gives an error and does not boot.

### Reproduction Steps ###

  1. Add ignition config to "user-data" file
  2. vagrant up
  3. Vagrant prints this error message:

```
There was an error loading a Vagrantfile. The file being loaded
and the error message are shown below. This is usually caused by
a syntax error.

Path: /Users/larstobi/coreos-vagrant/Vagrantfile
Line number: 0
Message: Psych::SyntaxError: (<unknown>): did not find expected key while parsing a block mapping at line 1 column 3
```

Which is because a line in config.rb tries to parse a Yaml file with cloud-config instead of a JSON file with ignition config.

### Other Information ###

**Comment by larstobi** [comment]
If I comment out all the yaml stuff in config.rb and just place the ignition config in the user-data file, the user-data file is successfully copied into the VM at /var/lib/coreos-vagrant/vagrantfile-user-data but the ignition config isn't being read by Container Linux.

From journalctl:

```
Mar 30 09:09:44 localhost ignition[227]: Ignition v0.13.0
Mar 30 09:09:44 localhost ignition[227]: parsed url from cmdline: ""
Mar 30 09:09:44 localhost ignition[227]: no config URL provided
Mar 30 09:09:44 localhost ignition[227]: noop provider fetching empty config
Mar 30 09:09:44 localhost ignition[227]: failed to fetch config: not a config (empty)
Mar 30 09:09:44 localhost ignition[227]: not a config (empty): ignoring user-provided config
```

**Comment by crawford** [comment]
Ignition runs long before that config file is written by Vagrant. In order for this to work, Ignition will need to support VirtualBox, which requires modifying the hypervisor (https://github.com/coreos/bugs/issues/1573). Once this is complete, we can turn on support in Vagrant for both VirtualBox and VMware.

In the mean time, VMware will work fine if you pass the configuration via the guest variables.

**Comment by robszumski** [comment]
Lack of Ignition support blocks updating our Kubernetes Vagrant to 1.6: https://github.com/coreos/coreos-kubernetes/issues/863#issuecomment-293712555

**Comment by crawford** [comment]
Vagrant is not an officially supported platform of Container Linux. Our Kubernetes support will need to be _heavily_ modified to support that platform or we wait until Ignition supports VirtualBox (not a priority right now).

**Comment by matjohn2** [comment]
Stumbled upon this thread while having the same issue converting a user-data file to ignition. Development env is Vagrant+Virtualbox, eventual env will be bare metal.

@crawford in regards to your comment:
> Vagrant is not an officially supported platform of Container Linux. Our Kubernetes support will need to be heavily modified to support that platform.

Where does that leave this document? https://coreos.com/kubernetes/docs/latest/kubernetes-on-vagrant-single.html 

Should this page not exist anymore, or should I not be doing anything like it for single node k8s+coreos development? If so, what is the suggested best practise for developing (when the eventual env will be bare metal appliances?).

Appreciate any comments/direction.

Thanks.

**Comment by dghubble** [comment]
His comment was referring to officially supported vs community supported platforms. No need to delete the vagrant document. Both coreos/coreos-kubernetes and kubernetes-incubator/bootkube provide Vagrant/Virtualbox local development flows, but neither allows using Ignition yet.

CoreOS Matchbox supports bare-metal provisioning via Container Linux config (and raw Ignition). You can find the "local QEMU/KVM" development environment and tutorials (designed to be as similar to production as possible) at https://coreos.com/matchbox/docs/latest/.

**Comment by matjohn2** [comment]
@dghubble Thanks for the info and the speedy response 👍

**Comment by delsvr** [comment]
Are there any updates here? I believe coreos/vagrant-ignition has a pull request specifically for supporting the vmware provider (https://github.com/coreos/vagrant-ignition/pull/5). Is the only thing stopping this a review?

**Comment by crawford** [comment]
@bgilbert Would you mind taking a look at this? The referenced PR looks reasonable (though, the hard-coded interface names make me nervous).

**Comment by bgilbert** [comment]
@delsvr Our Vagrant support needs work, both on the VMware and VirtualBox sides, and in addition to what's in coreos/vagrant-ignition#5.  Unfortunately that work isn't a priority right now.

**Comment by bgilbert** [comment]
Thank you for reporting this issue. Unfortunately, we don't think we'll end up addressing it in Container Linux.

We're now working on [Fedora CoreOS](https://coreos.fedoraproject.org/), the successor to Container Linux, and we expect most major development to occur there instead. Meanwhile, Container Linux will be fully maintained into 2020 but won't see many new features. We appreciate your taking the time to report this issue and we're sorry that we won't be able to address it.
